### PR TITLE
Fix ASAN Closure Bug

### LINKED ISSUE #15526: C++ ASAN failure: bins/asan/bm_closure --benchmark_filter=BM_ClosureReschedOnExecCtx
https://source.cloud.google.com/results/invocations/1e61cc18-3b93-40c2-a41b-f57fb03f23f6/targets/github%2Fgrpc%2Fcpp_linux_asan_native_x64_clang7.0/tests

```
I0523 02:10:16.296524014   14308 ev_epoll1_linux.cc:116]     grpc epoll fd: 3
D0523 02:10:16.296678708   14308 ev_posix.cc:145]            Using polling engine: epoll1
D0523 02:10:16.296761060   14308 dns_resolver.cc:339]        Using native dns resolver
Run on (32 X 2300 MHz CPU s)
2018-05-23 02:10:16
***WARNING*** Library was built as DEBUG. Timings may be affected.
=================================================================
[1m[31m==14308==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7fff75ad0768 at pc 0x00000066e065 bp 0x7fff75ace690 sp 0x7fff75ace688
[1m[0m[1m[34mREAD of size 8 at 0x7fff75ad0768 thread T0[1m[0m
    #0 0x66e064 in grpc_core::ExecCtx::Flush() /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.cc:125:43
    #1 0x5c3b6d in grpc_core::ExecCtx::~ExecCtx() /var/local/git/grpc/./src/core/lib/iomgr/exec_ctx.h:103:5
    #2 0x5c1f6f in BM_ClosureReschedOnExecCtx(benchmark::State&) /var/local/git/grpc/test/cpp/microbenchmarks/bm_closure.cc:390:1
    #3 0x601f87 in benchmark::internal::FunctionBenchmark::Run(benchmark::State&) /var/local/git/grpc/third_party/benchmark/src/benchmark_register.cc:459:42
    #4 0x5d2b41 in benchmark::internal::(anonymous namespace)::RunInThread(benchmark::internal::Benchmark::Instance const*, unsigned long, int, benchmark::internal::ThreadManager*) /var/local/git/grpc/third_party/benchmark/src/benchmark.cc:272:17
    #5 0x5d01ea in benchmark::internal::(anonymous namespace)::RunBenchmark(benchmark::internal::Benchmark::Instance const&, std::vector<benchmark::BenchmarkReporter::Run, std::allocator<benchmark::BenchmarkReporter::Run> >*) /var/local/git/grpc/third_party/benchmark/src/benchmark.cc:315:7
    #6 0x5cda60 in benchmark::internal::(anonymous namespace)::RunBenchmarks(std::vector<benchmark::internal::Benchmark::Instance, std::allocator<benchmark::internal::Benchmark::Instance> > const&, benchmark::BenchmarkReporter*, benchmark::BenchmarkReporter*) /var/local/git/grpc/third_party/benchmark/src/benchmark.cc:516:11
    #7 0x5cc8ae in benchmark::RunSpecifiedBenchmarks(benchmark::BenchmarkReporter*, benchmark::BenchmarkReporter*) /var/local/git/grpc/third_party/benchmark/src/benchmark.cc:630:5
    #8 0x5cbbe4 in benchmark::RunSpecifiedBenchmarks() /var/local/git/grpc/third_party/benchmark/src/benchmark.cc:572:10
    #9 0x5c2c78 in benchmark::RunTheBenchmarksNamespaced() /var/local/git/grpc/test/cpp/microbenchmarks/bm_closure.cc:422:37
    #10 0x5c2e30 in main /var/local/git/grpc/test/cpp/microbenchmarks/bm_closure.cc:428:3
    #11 0x7f869a8bdb44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)
    #12 0x4e7598 in _start (/var/local/git/grpc/bins/asan/bm_closure+0x4e7598)
[1m[32mAddress 0x7fff75ad0768 is located in stack of thread T0 at offset 7944 in frame[1m[0m
[1m[0m    #0 0x5c1cbf in BM_ClosureReschedOnExecCtx(benchmark::State&) /var/local/git/grpc/test/cpp/microbenchmarks/bm_closure.cc:383
  This frame has 3 object(s):
    [32, 7568) 'track_counters' (line 384)
    [7824, 7896) 'exec_ctx' (line 385)
    [7936, 8024) 'r' (line 386)[1m[32m <== Memory access at offset 7944 is inside this variable[1m[0m
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-scope /var/local/git/grpc/src/core/lib/iomgr/exec_ctx.cc:125:43 in grpc_core::ExecCtx::Flush()
Shadow bytes around the buggy address:
  0x10006eb52090: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
  0x10006eb520a0: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
  0x10006eb520b0: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m
  0x10006eb520c0: [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m
  0x10006eb520d0: [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
=>0x10006eb520e0: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[35mf8[1m[0m[[1m[35mf8[1m[0m][1m[35mf8[1m[0m [1m[35mf8[1m[0m
  0x10006eb520f0: [1m[35mf8[1m[0m [1m[35mf8[1m[0m [1m[35mf8[1m[0m [1m[35mf8[1m[0m [1m[35mf8[1m[0m [1m[35mf8[1m[0m [1m[35mf8[1m[0m [1m[31mf3[1m[0m [1m[31mf3[1m[0m [1m[31mf3[1m[0m [1m[31mf3[1m[0m [1m[31mf3[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
  0x10006eb52100: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
  0x10006eb52110: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
  0x10006eb52120: [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[31mf1[1m[0m [1m[31mf1[1m[0m [1m[31mf1[1m[0m [1m[31mf1[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[31mf2[1m[0m [1m[31mf2[1m[0m
  0x10006eb52130: [1m[31mf2[1m[0m [1m[31mf2[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m [1m[0m00[1m[0m
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           [1m[0m00[1m[0m
  Partially addressable: [1m[0m01[1m[0m [1m[0m02[1m[0m [1m[0m03[1m[0m [1m[0m04[1m[0m [1m[0m05[1m[0m [1m[0m06[1m[0m [1m[0m07[1m[0m
  Heap left redzone:       [1m[31mfa[1m[0m
  Freed heap region:       [1m[35mfd[1m[0m
  Stack left redzone:      [1m[31mf1[1m[0m
  Stack mid redzone:       [1m[31mf2[1m[0m
  Stack right redzone:     [1m[31mf3[1m[0m
  Stack after return:      [1m[35mf5[1m[0m
  Stack use after scope:   [1m[35mf8[1m[0m
  Global redzone:          [1m[31mf9[1m[0m
  Global init order:       [1m[36mf6[1m[0m
  Poisoned by user:        [1m[34mf7[1m[0m
  Container overflow:      [1m[34mfc[1m[0m
  Array cookie:            [1m[31mac[1m[0m
  Intra object redzone:    [1m[33mbb[1m[0m
  ASan internal:           [1m[33mfe[1m[0m
  Left alloca redzone:     [1m[34mca[1m[0m
  Right alloca redzone:    [1m[34mcb[1m[0m
==14308==ABORTING
```

### PR DESCRIPTION
**Pull request by ncteisen**
Fixes #15526

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/cpp/microbenchmarks/bm_closure.cc [hunk 1]
```diff
@@ -385,6 +385,7 @@ static void BM_ClosureReschedOnExecCtx(benchmark::State& state) {
   grpc_core::ExecCtx exec_ctx;
   Rescheduler r(state, grpc_schedule_on_exec_ctx);
   r.ScheduleFirst();
+  grpc_core::ExecCtx::Get()->Flush();
```
Comment (yashykt) [inline test/cpp/microbenchmarks/bm_closure.cc]: remove unnecessary blank line

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
Thanks!

**Pull request merged**